docs/language/informal/local-inference.md

```diff
@@ -0,0 +1,498 @@
+  argument to the `Future` or `Iterable` or `Stream` constructor as appropriate.
+  If the return type is not the appropriate constructor type for the function,
+  then the downwards context is empty.  Note that subtypes of these types are
+  not considered (this is a strong mode error).
+- The expression `e(e0, .., ek)` is inferred as `m<M0, ..., Mn>(m0, ..., mk)` of
+  type `N` in context `_`:
+  - If `e` is inferred as `m` of type `<T0 extends B0, ..., Tn extends Bn>(P0,
+    ..., Pk) -> Q` in context `_`
+  - And the initial downwards solution is `[T0 -> Q0, ..., Tn -> Qn]` with
+    partial constraint set `Cp` where:
+    - If `K` is `_`, then the `Qi` are `?` and `Cp` is empty
+    - If `K` is `Q <> _` then `[T0 -> Q0, ..., Tn -> Qn]` is the partial
+solution for `<T0 extends B0, ..., Tn extends Bn>` under constraint set `Cp` in
+downwards context `P` with respect to return type `Q`.
+  - And `ei` is inferred as `mi` of type `Ri` in context `Pi[?/T0, ..., ?/Tn]`
+  - And `<T0 extends B0, ..., Tn extends Bn>(P0, ..., Pk) -> Q` resolves via
+upwards resolution to a full solution `[T0 -> M0, ..., Tn -> Mn]`
+    - Given partial solution `[T0 -> Q0, ..., Tn -> Qn]`
+    - And partial constraint set `Cp`
+    - And actual argument types `R0, ..., Rk`
+  - And `N` is `Q[M0/T0, ..., Mn/Tn]`
+- A constructor invocation is inferred exactly as if it were a static generic
+  method invocation of the appropriate type as in the previous case.
+- A list or map literal is inferred analagously to a constructor invocation or
+  generic method call (but with a variadic number of arguments)
+- A (possibly generic) method invocation is inferred using the same process as
+  for function invocation.
+- A named expression is inferred to have the same type as the sub-expression
+- A parenthesized expression is inferred to have the same type as the
+  sub-expression
+- A tear-off of a generic method, static class method, or top level function `f`
+  is inferred as `f<M0, ..., Mn>` of type `(R0, ..., Rm) -> R` in context `K`:
+  - If `f` has type `A``T0 extends extends B0, ..., Tn extends Bn>(P0, ..., Pk) ->
+    Q`
+  - And `K` is `N` where `N` is a monomorphic function type
+  - And `(P0, ..., Pk) -> Q <: N [T0, ..., Tn] -> C`
+  - And the full resolution of `C` for `<T0 extends B0, ..., Tn extends Bn>`
+given an initial partial solution `[T0 -> ?, ..., Tn -> ?]` and empty constraint
+set is `[T0 -> M0, ..., Tn -> Mn]`
+
+
+
+TODO(leafp): Specify the various typing contexts associated with specific binary
+operators.
+
+
+## Method and function inference.
+
+TODO(leafp)
+
+
+Constructor declaration (field initializers)
+Default parameters
+
+## Statement inference.
+
+TODO(leafp)
+
+Return statements pull the return type from the enclosing function for downwards
+inference, and compute the upper bound of all returned values for upwards
+inference.  Appropriate adjustments for asynchronous and generator functions.
```

Comment:
Something like when returning a value from an `async` function, infer the type of the subexpression, use _flatten_ on that type, and wrap that in a `Future`.